EIPS/eip-758.md

```diff
@@ -0,0 +1,96 @@
+## Preamble
+
+    EIP: 758
+    Title: Subscriptions and filters for transaction return data
+    Author: Jack Peterson <jack@tinybike.net>
+    Type: Standard Track
+    Category: Interface
+    Status: Draft
+    Created: 2017-11-09
+
+
+## Simple Summary
+Provide a way for external callers to access the return data of functions executed during Ethereum transactions.
+
+## Abstract
+When a new transaction is submitted successfully to an Ethereum node, the node responds with the transaction's hash.  If the transaction involved the execution of a contract function that returns data, the data is discarded.  If the return data is state-dependent, which is common, there is no straightforward way for the caller to access or compute the return data.  This EIP proposes that callers should be able to subscribe to (or poll for) the return data of their transactions.  The Ethereum node then sends the return data to the caller when the transactions are sealed.
+
+## Motivation
+External callers presently have no way of accessing return data from Ethereum, if the function was executed via `eth_sendTransaction` or `eth_sendRawTransaction` RPC request.  Access to function return data is in many cases a desirable feature.  Making return data available to external callers also addresses the inconsistency between internal callers, which have access to return data within the context of the transaction, and external callers, which do not.  Presently, a common workaround is to log the return data, which is bad for several reasons: it contributes to chain bloat, imposes additional gas costs on the caller, and can result in unused logs being written if the externally called function involves other (internal) function calls that log their return data.
+
+## Specification
+
+### Subscription
+A caller who wants to be notified of return data for their transactions sends an `eth_subscribe` RPC request with the parameter `"returnData"`:
+
+```json
+{"jsonrpc": "2.0", "id": 1, "method": "eth_subscribe", "params": ["returnData"]}
+```
+
+The Ethereum node responds with a subscription ID:
+
+```json
+{"jsonrpc": "2.0", "id": 1, "result": "0x00000000000000000000000000000b0b"}
+```
+
+The caller submits a transaction via `eth_sendTransaction` or `eth_sendRawTransaction` RPC request which has the transaction hash `"0x00000000000000000000000000000000000000000000000000000000deadbeef"`.  When the transaction is sealed (mined), the Ethereum node computes the return value (`"0x000000000000000000000000000000000000000000000000000000000000002a"`) and pushes a notification to the caller:
```

Comment:
Does this notification happen for
a. every transaction in every block?
b. transactions submitted from the node?
c. transactions submitted from the node during subscription?
d. the first transaction submitted after the subscription starts?

Reply:
My intention was somewhere between (b) and (c): after `returnData` subscription, all transactions submitted from the node that are mined _after_ the subscription begins would be included.  (So, if I submit a transaction, then subscribe while the transaction is sitting in the txpool, I would receive the `returnData` notification when the transaction is mined.)

FWIW, I can see an argument for (a) or (d).  The reason I shied away from (a) is because I wanted to constrain the amount of extra computational load this could potentially place on nodes.  I shied away from (d) because I _think_ the most common use-case will be to listen for the return data for all of your own transactions, and having to submit a new eth_subscribe-to-returnData request to the node seems a bit onerous to me.  (If the more common use-case turns out to be subscribing to particular transactions, it seems to me the way to do that would be to allow users to pass in a transaction hash to `eth_subscribe`.  To be clear, that's _not_ the intent of this proposal, just mentioning this as a possible follow-on, if this EIP is accepted.)